tls/extensions/s2n_server_key_share.c

```diff
@@ -127,8 +129,8 @@ int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn)
     /* If we are responding to a retry request then we don't have a valid
      * curve from the client. Just return 0 so a selected group will be
      * chosen for the key share. */
-    if (s2n_is_hello_retry_required(conn)) {
-        return 0;
+    if (s2n_is_hello_retry_message(conn) && s2n_is_hello_retry_valid(conn)) {
```

Comment:
Why do we need to continuously check that the hello_retry is valid? Isn't it sufficient to just check once before setting the handshake type?

Reply:
We would require this check,  since after receiving the ClientHello2, while sending the ServerHello response, the handshake type is still set to`HELLO_RETRY_REQUEST` and this check will pass when we expect it to fail. This `if` needs to be true only when the message type is `HELLO_RETRY_MSG` and if, client_version and server_version is TLS13 and the server_random is set to the hrr_random value.